File: answers/leetcode/Remove Duplicates from Sorted List II/Remove Duplicates from Sorted List II.cpp
//@result 166 / 166 test cases passed. Status: Accepted Runtime: 12 ms Submitted: 2 minutes ago You are here! Your runtime beats 8.24% of cpp submissions.

// 2014-12-24
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
		ListNode *dumb = new ListNode(INT_MAX);
		dumb->next = head;
		ListNode *current = dumb;
		while (current) {
			if (current->next && current->next->next && current->next->val == current->next->next->val) {
				ListNode *new_val = current->next;
				while (new_val && new_val->val == current->next->val) {
					new_val = new_val->next;
				}
				current->next = new_val;
			} else {
				current = current->next;
			}
		}
		return dumb->next;
    }
};

File: answers/leetcode/Remove Duplicates from Sorted List II/Remove Duplicates from Sorted List II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        int count = 0;
        ListNode left = dummy;
        ListNode right = head;
        while (true) {
            if (right != null && left.next.val == right.val) {
                count++;
            } else if (count > 1) {
                count = 1;
                left.next = right;
            } else {
                count = 1;
                left = left.next;
            }
            if (right == null) {
                break;
            }
            right = right.next;
        }
        return dummy.next;
    }
}